### PR TITLE
Adding a new in memory database to core-data

### PR DESCRIPTION
**Pull request by feclare**
This PR adds a new database type to core-data that does not use any backend, only keeps the data in memory while it is running and data is lost when core-data stops. It is usefull while developing or demos.

It also add some generic unittest for DBClient, that could be used to test other implementation. Something similar to what is done in https://github.com/edgexfoundry/edgex-go/blob/master/export/client/clients/mongo-client_test.go

### CODE REVIEW

**Comment by tsconn23** [comment]
recheck

**Comment by tsconn23** [comment]
Please hold off on any more merges until tomorrow May, 22nd CST hours.

File: core/data/clients/memory-database_test.go [hunk 1]
```diff
@@ -0,0 +1,612 @@
...
+	}
+}
+
+func testDB(t *testing.T, db DBClient) {
+	testDBReadings(t, db)
+	testDBEvents(t, db)
+	testDBValueDescriptors(t, db)
+
+	db.CloseSession()
+	db.CloseSession()
```
Comment (jduranf) [inline core/data/clients/memory-database_test.go]: This line should be removed

Comment (feclare) [reply core/data/clients/memory-database_test.go]: I'm calling the ```CloseSession``` a second time to test that there is no panic when closing a db already closed. I will add a comment.

Comment (jduranf) [reply core/data/clients/memory-database_test.go]: ok! thanks for the clarification

**Comment by tsconn23** [comment]
recheck

**Pull request merged**